File: swift/Sources/AlloyAudio/DSP/Effects/Compressor.swift
import Foundation

/// Stereo-linked feed-forward compressor — a single max(|L|, |R|) detector
/// drives both channels identically (link), so a loud transient on one
/// channel ducks both together instead of skewing the stereo image. The
/// detector's one-pole attack/release smoothing runs every sample (full
/// rate); the log/pow-heavy gain computation runs once per
/// EffectConstants.controlInterval samples (control rate) and is held
/// constant across the tick — same two-rate philosophy as the phaser's
/// swept coefficient (see Phaser.swift). Twin of web
/// src/dsp/effects/compressor.ts (canonical).
public final class Compressor: EffectUnit {
    private let params: CompressorParams
    private var env = 0.0
    private var gain: Double
    private var sampleCounter = 0
    private let attackCoef: Double
    private let releaseCoef: Double

    public init(params: CompressorParams, sampleRate: Double) {
        self.params = params
        self.attackCoef = 1 - exp(-1 / ((params.attackMs / 1000) * sampleRate))
        self.releaseCoef = 1 - exp(-1 / ((params.releaseMs / 1000) * sampleRate))
        self.gain = pow(10, params.makeupDb / 20)
    }

    public func reset() {
        env = 0
        gain = pow(10, params.makeupDb / 20)
        sampleCounter = 0
    }

    public func process(left: inout [Float], right: inout [Float], frames: Int) {
        let thresholdDb = params.thresholdDb
        let ratio = params.ratio
        let makeupDb = params.makeupDb

        for i in 0..<frames {
            let l = Double(left[i])
            let r = Double(right[i])

            let d = max(abs(l), abs(r))
            env += (d > env ? attackCoef : releaseCoef) * (d - env)

            if sampleCounter % EffectConstants.controlInterval == 0 {
                let envDb = 20 * log10(max(env, 1e-6))
                let over = max(0, envDb - thresholdDb)
                let reductionDb = over * (1 - 1 / ratio)
                gain = pow(10, (makeupDb - reductionDb) / 20)
            }

            left[i] = Float(l * gain)
            right[i] = Float(r * gain)

            sampleCounter += 1
        }
    }
}

File: web/packages/alloy-audio/src/dsp/effects/compressor.ts
// Stereo-linked feed-forward compressor — a single max(|L|, |R|) detector
// drives both channels identically (link), so a loud transient on one
// channel ducks both together instead of skewing the stereo image. The
// detector's one-pole attack/release smoothing runs every sample (full
// rate); the log/pow-heavy gain computation runs once per
// EFFECT_CONTROL_INTERVAL samples (control rate) and is held constant
// across the tick — same two-rate philosophy as the phaser's swept
// coefficient (see phaser.ts). Twin: Compressor.swift.

import { EFFECT_CONTROL_INTERVAL, type CompressorParams, type EffectUnit } from './effect-types.js';

export class Compressor implements EffectUnit {
  private env = 0;
  private gain: number;
  private sampleCounter = 0;
  private readonly attackCoef: number;
  private readonly releaseCoef: number;

  constructor(
    private readonly params: CompressorParams,
    sampleRate: number,
  ) {
    this.attackCoef = 1 - Math.exp(-1 / ((params.attackMs / 1000) * sampleRate));
    this.releaseCoef = 1 - Math.exp(-1 / ((params.releaseMs / 1000) * sampleRate));
    this.gain = 10 ** (params.makeupDb / 20);
  }

  reset(): void {
    this.env = 0;
    this.gain = 10 ** (this.params.makeupDb / 20);
    this.sampleCounter = 0;
  }

  process(left: Float32Array, right: Float32Array, frames: number): void {
    const { thresholdDb, ratio, makeupDb } = this.params;

    for (let i = 0; i < frames; i++) {
      const l = left[i];
      const r = right[i];

      const d = Math.max(Math.abs(l), Math.abs(r));
      this.env += (d > this.env ? this.attackCoef : this.releaseCoef) * (d - this.env);

      if (this.sampleCounter % EFFECT_CONTROL_INTERVAL === 0) {
        const envDb = 20 * Math.log10(Math.max(this.env, 1e-6));
        const over = Math.max(0, envDb - thresholdDb);
        const reductionDb = over * (1 - 1 / ratio);
        this.gain = 10 ** ((makeupDb - reductionDb) / 20);
      }

      left[i] = l * this.gain;
      right[i] = r * this.gain;

      this.sampleCounter++;
    }
  }
}
